File: project_euler/Problem093/Problem093.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem093() {
    signs := []byte{'+', '-', '*', '/'}
    operations := make([][]byte, 0)
    getOperations(signs, &operations)

    elements := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
    auxiliary := make([]float64, 4)
    result := Arithmetic{0, 0}
    calculations := make(map[int]bool)
    combination(elements, 0, 4, auxiliary, 0, operations, &result, calculations)
    fmt.Println(result.value)
}

func getOperations(signs []byte, operations *[][]byte) {
    auxiliary := make([]byte, 3)
    search(signs, 3, operations, auxiliary, 0)
}

func combination(elements []int, index int, count int, auxiliary []float64, size int, operations [][]byte, pivot *Arithmetic, calculations map[int]bool) {
    if count == size {
        calculations = make(map[int]bool)
        permutation(auxiliary, 0, count, operations, calculations)
        border := 1
        for calculations[border] {
            border++
        }
        border--
        if pivot.border < border {
            pivot.border = border
            pivot.value = 0
            for _, value := range auxiliary {
                pivot.value = pivot.value * 10 + int(value)
            }
        }
        return
    }

    for i, value := range elements[index:] {
        auxiliary[size] = float64(value)
        size++
        combination(elements, index + i + 1, count, auxiliary, size, operations, pivot, calculations)
        size--
    }
}

func search(signs []byte, count int, operations *[][]byte, auxiliary []byte, size int) {
    if count == size {
        *operations = append(*operations, make([]byte, count))
        copy((*operations)[len(*operations) - 1], auxiliary)
        return
    }

    for _, value := range signs {
        auxiliary[size] = value
        size++
        search(signs, count, operations, auxiliary, size)
        size--
    }
}

func permutation(elements []float64, index int, length int, operations [][]byte, calculations map[int]bool) {
    if index == length {
        for _, operation := range operations {
            addCalculations(calculate(operation[2], calculate(operation[1], calculate(operation[0], elements[0], elements[1]), elements[2]), elements[3]), calculations)
            addCalculations(calculate(operation[1], calculate(operation[0], elements[0], elements[1]), calculate(operation[2], elements[2], elements[3])), calculations)
            addCalculations(calculate(operation[2], calculate(operation[0], elements[0], calculate(operation[1], elements[1], elements[2])), elements[3]), calculations)
            addCalculations(calculate(operation[0], elements[0], calculate(operation[2], calculate(operation[1], elements[1], elements[2]), elements[3])), calculations)
            addCalculations(calculate(operation[0], elements[0], calculate(operation[1], elements[1], calculate(operation[2], elements[2], elements[3]))), calculations)
            addCalculations(calculate(operation[0], elements[0], calculate(operation[1], elements[1], calculate(operation[2], elements[2], elements[3]))), calculations)
        }
        return
    }

    permutation(elements, index + 1, length, operations, calculations)
    for i := index + 1; i < length; i++ {
        elements[index], elements[i] = elements[i], elements[index]
        permutation(elements, index + 1, length, operations, calculations)
        elements[index], elements[i] = elements[i], elements[index]
    }
}

func addCalculations(value float64, calculations map[int]bool) {
    if value > 0 && math.Abs(math.Floor(value + 0.5) - value) < 0.00000001 {
        calculations[int(value)] = true
    }
}

func calculate(sign byte, i float64, j float64) float64 {
    if '+' == sign {
        return i + j
    }
    if '-' == sign {
        return i - j
    }
    if '*' == sign {
        return i * j
    }
    if math.Abs(j) < 0.00000001 {
        return -1
    }
    return i / j
}

type Arithmetic struct {
    border int
    value int
}

File: project_euler/Problem093/Problem093.py
import math
from typing import List

class Problem093:
    def solution(self) -> None:
        signs, operations = ['+', '-', '*', '/'], []
        self.__get_operations(signs, operations)

        elements, auxiliary, result, calculations = [1, 2, 3, 4, 5, 6, 7, 8, 9], [0 for _ in range(4)], self.__Arithmetic(0, 0), set()
        self.__combination(elements, 0, 4, auxiliary, 0, operations, result, calculations)
        print(result.value)

    def __get_operations(self, signs: List[str], operations: List[List[str]]) -> None:
        auxiliary = [' ' for _ in range(3)]
        self.__search(signs, 3, operations, auxiliary, 0)

    def __combination(self, elements: List[int], index: int, count: int, auxiliary: List[int], size: int, operations: List[List[str]], pivot: '__Arithmetic', calculations: set[int]) -> None:
        if count == size:
            calculations.clear()
            self.__permutation(auxiliary, 0, count, operations, calculations)
            border = 1
            while border in calculations:
                border += 1
            border -= 1
            if pivot.border < border:
                pivot.border, pivot.value = border, 0
                for value in auxiliary:
                    pivot.value = pivot.value * 10 + value
            return

        for i, value in enumerate(elements[index:]):
            auxiliary[size] = value
            size += 1
            self.__combination(elements, index + i + 1, count, auxiliary, size, operations, pivot, calculations)
            size -= 1

    def __search(self, signs: List[str], count: int, operations: List[List[str]], auxiliary: List[str], size: int) -> None:
        if count == size:
            operations.append(list(auxiliary))
            return

        for value in signs:
            auxiliary[size] = value
            size += 1
            self.__search(signs, count, operations, auxiliary, size)
            size -= 1

    def __permutation(self, elements: List[int], index: int, length: int, operations: List[List[str]], calculations: set[int]) -> None:
        if index == length:
            for operation in operations:
                self.__add_calculations(self.__calculate(operation[2], self.__calculate(operation[1], self.__calculate(operation[0], elements[0], elements[1]), elements[2]), elements[3]), calculations)
                self.__add_calculations(self.__calculate(operation[1], self.__calculate(operation[0], elements[0], elements[1]), self.__calculate(operation[2], elements[2], elements[3])), calculations)
                self.__add_calculations(self.__calculate(operation[2], self.__calculate(operation[0], elements[0], self.__calculate(operation[1], elements[1], elements[2])), elements[3]), calculations)
                self.__add_calculations(self.__calculate(operation[0], elements[0], self.__calculate(operation[2], self.__calculate(operation[1], elements[1], elements[2]), elements[3])), calculations)
                self.__add_calculations(self.__calculate(operation[0], elements[0], self.__calculate(operation[1], elements[1], self.__calculate(operation[2], elements[2], elements[3]))), calculations)
            return

        self.__permutation(elements, index + 1, length, operations, calculations)
        for i in range(index + 1, length):
            elements[index], elements[i] = elements[i], elements[index]
            self.__permutation(elements, index + 1, length, operations, calculations)
            elements[index], elements[i] = elements[i], elements[index]

    def __add_calculations(self, value: float, calculations: set[int]) -> None:
        if value > 0 and abs(math.floor(value + 0.5) - value) < 0.00000001:
            calculations.add(int(value))

    def __calculate(self, sign: str, i: float, j: float) -> float:
        if '+' == sign:
            return i + j
        if '-' == sign:
            return i - j
        if '*' == sign:
            return i * j
        if abs(j) < 0.00000001:
            return -1
        return i / j

    class __Arithmetic:
        def __init__(self, border: int, value: int) -> None:
            self.border, self.value = border, value
